chia/full_node/subscriptions.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+from typing import Dict, List, Optional, Set
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+
+
+def _get_or_add_set(c: Dict[bytes32, Set[bytes32]], key: bytes32) -> Set[bytes32]:
+    ret: Optional[Set[bytes32]] = c.get(key)
+    if ret is not None:
+        return ret
+
+    ret2: Set[bytes32] = set()
+    c[key] = ret2
+    return ret2
+
+
+# The PeerSubscriptions class is essentially a multi-index container. It can be
+# indexed by peer_id, coin_id and puzzle_hash.
+class PeerSubscriptions:
+    # TODO: use NewType all over to describe these various uses of the same types
+    # Puzzle Hash : Set[Peer ID]
+    _coin_subscriptions: Dict[bytes32, Set[bytes32]]
+    # Puzzle Hash : Set[Peer ID]
+    _ph_subscriptions: Dict[bytes32, Set[bytes32]]
+    # Peer ID: Set[Coin ids]
+    _peer_coin_ids: Dict[bytes32, Set[bytes32]]
+    # Peer ID: Set[puzzle_hash]
+    _peer_puzzle_hash: Dict[bytes32, Set[bytes32]]
+    # Peer ID: subscription count
+    _peer_sub_counter: Dict[bytes32, int]
+
+    def has_ph_sub(self, ph: bytes32) -> bool:
+        return ph in self._ph_subscriptions
+
+    def has_coin_sub(self, coin_id: bytes32) -> bool:
+        return coin_id in self._coin_subscriptions
+
+    def __init__(self) -> None:
+        self._coin_subscriptions = {}
+        self._ph_subscriptions = {}
+        self._peer_coin_ids = {}
+        self._peer_puzzle_hash = {}
+        self._peer_sub_counter = {}
```

Comment:
Should be a dataclass.

Reply:
reading up on dataclasses a bit, it doesn't look like they're meant for this. It looks like `@dataclass` is meant for value types.

Reply:
It's not clear to me what the benefit is to add:
```
@dataclass(init=False, repr=False, eq=False, order=False, unsafe_hash=True, match_args=False)
```

to the class.

Reply:
I'm not sure why there is a tendency to think that a clear and consistently structured class is only useful for 'small' or 'trivial' or 'attribute only' classes.  Not that the doc is definitively correct or anything, but since I wrote up my reasons there, here's a link https://github.com/Chia-Network/chia-blockchain/blob/b8068a856d8e2ccd6660f5f6488182e3c1589be8/PRETTY_GOOD_PRACTICES.md#dataclasses.

Why would you write that big huge dataclass decorator that defeats the basic benefit of a dataclass by still using your hand coded init that duplicates the list of hinted attributes?  I'll add a separate comment with a suggestion to show the nominal starting point I would use.